src/core/Track.cpp

```diff
@@ -889,76 +896,86 @@ void TrackContentObjectView::mouseMoveEvent( QMouseEvent * me )
 		{
 			TrackContentObjectView * tcov =
 				dynamic_cast<TrackContentObjectView *>( *it );
-			if( tcov == NULL )
-			{
-				continue;
-			}
-			TrackContentObject * tco = tcov->m_tco;
-			tcos.push_back( tco );
-			smallestPos = qMin<int>( smallestPos,
-					(int)tco->startPosition() + dtick );
-		}
-		dtick -= smallestPos;
-		if( snap )
-		{
-			dtick = dtick.toAbsoluteTact(); // round toward 0
+			if( tcov == NULL ) { continue; }
+			tcos.push_back( tcov->m_tco );
+			int index = std::distance( so.begin(), it );
+			leftmost = min (leftmost, m_initialOffsets[index].getTicks() );
 		}
+		// Make sure the leftmost clip doesn't get moved to a negative position
+		if ( newPos.getTicks() + leftmost < 0 ) { newPos = -leftmost; }
+
 		for( QVector<TrackContentObject *>::iterator it = tcos.begin();
 							it != tcos.end(); ++it )
 		{
-			( *it )->movePosition( ( *it )->startPosition() + dtick );
+			int index = std::distance( tcos.begin(), it );
+			( *it )->movePosition( newPos + m_initialOffsets[index] );
 		}
 	}
 	else if( m_action == Resize || m_action == ResizeLeft )
 	{
 		if( m_action == Resize )
 		{
-			MidiTime t = qMax( MidiTime::ticksPerTact() / 16, static_cast<int>( me->x() * MidiTime::ticksPerTact() / ppt ) );
-			if( ! ( me->modifiers() & Qt::ControlModifier ) && me->button() == Qt::NoButton )
+			MidiTime l = qMax( MidiTime::ticksPerTact() / 16, static_cast<int>( me->x() * MidiTime::ticksPerTact() / ppt ) );
+			float snapSize = gui->songEditor()->m_editor->getSnapSize();
+			//When snapping, avoid making the new TCO tiny
+			MidiTime minLength = MidiTime( MidiTime::ticksPerTact() / 16 );
+
+			// If the user is holding alt, or pressed ctrl after beginning the drag, don't quantize
+			if (    me->button() != Qt::NoButton
+				|| (me->modifiers() & Qt::ControlModifier)
+				|| (me->modifiers() & Qt::AltModifier)    )
 			{
-				t = qMax<int>( MidiTime::ticksPerTact(), t.toNearestTact() );
+				//If the user manually makes the TCO tiny, we ĺet them
+				minLength = MidiTime( 1 );
+				// We want to preserve this adjusted offset,
+				// even if the user switches to snapping
+				setInitialPos( m_initialMousePos );
+			}
+			else if ( me->modifiers() & Qt::ShiftModifier )
+			{	// If shift is held, quantize clip's end position
+				MidiTime end = MidiTime( m_initialTCOPos + l ).quantize( snapSize );
+				MidiTime min = m_initialTCOPos.quantize( snapSize );
+				if ( min < m_initialTCOPos )
+				{ min += MidiTime( MidiTime::ticksPerTact() * snapSize ); }
+				l = qMax<int>(min - m_initialTCOPos, end - m_initialTCOPos);
 			}
-			m_tco->changeLength( t );
+			else
+			{	// Otherwise, resize in fixed increments
+				MidiTime initialLength = m_initialTCOEnd - m_initialTCOPos;
+				MidiTime offset = MidiTime( l - initialLength ).quantize( snapSize );
+				MidiTime min = MidiTime( initialLength % (int)(MidiTime::ticksPerTact() * snapSize) );
+				l = qMax<int>( min, initialLength + offset);
+			}
+			l = qMax<int>( minLength, l );
```

Comment:
In some cases, part of the length which is out of quantization won't be preserved. For example:
1. Suppose there's a TCO whose length is 1 bar.
2. Shrink the TCO when the quantization size is 1 bar. Now the length of the TCO is 1/16 bars.
3. Release the mouse and extend the TCO again, it would be 1 and 1/16 bars instead of 1 bar.

Reply:
The length/position is always based on the initial value, or on the latest fine adjust, so I don't quite see how this is a bug. To me, releasing the mouse = applying the change. What's the alternative?

Reply:
What I meant was something like #3989. Related suggestion in the issue: https://github.com/LMMS/lmms/issues/3989#issuecomment-353770092

Reply:
I suppose I could/should change minLength to be ``note length modulo snap size``. I believe I set it to 1/16ths because at that size a clip can still be resized back up at 100% zoom. 1/8ths and smaller are a bit troublesome because they're very hard or impossible to size back up unless you zoom in.

Reply:
Latest commit ought to address this!

Reply:
Works great! :+1: